### PR TITLE
ansible/helm : align flags and deprecate the old ones

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description of the change:**
- Deprecated the flags which were removed in the Golang project in order to keep it better aligned with K8S
- Add the new flags
- Ensure that that metrics port is passed to the proxy. 

**Motivation for the change:**

- Reduce the complexities for we address https://github.com/kubernetes-sigs/kubebuilder/issues/2015
- Fix docs that were not updated.
- Keep Ansible/Helm/Go aligned (Align Helm/Ansible plugins with the changes made for Golang ( go/v3 ))
- Closes: Ansible/Helm providing the same flags as Go #4627

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/plugins/ansible/v1/scaffolds/internal/templates/config/kdefault/manager_auth_proxy_patch.go [hunk 1]
```diff
@@ -23,17 +23,16 @@ import (
 	"sigs.k8s.io/kubebuilder/v3/pkg/model/file"
 )
 
-var _ file.Template = &AuthProxyPatch{}
+var _ file.Template = &ManagerAuthProxyPatch{}
```
Comment (camilamacedo86) [inline internal/plugins/ansible/v1/scaffolds/internal/templates/config/kdefault/manager_auth_proxy_patch.go:26]: To keep the same names of upstream. Otherwise is harder yet to diff the files.

File: internal/ansible/flags/flag.go [hunk 2]
```diff
@@ -84,6 +86,12 @@ func (f *Flags) AddTo(flagSet *pflag.FlagSet) {
 		":8080",
 		"The address the metric endpoint binds to",
 	)
+	_ = flagSet.MarkDeprecated("metrics-addr", "Use the flag metrics-bind-address instead of")
```
Comment (estroz) [inline internal/ansible/flags/flag.go:89]:
```suggestion
	_ = flagSet.MarkDeprecated("metrics-addr", "use --metrics-bind-address instead")
```

File: internal/ansible/flags/flag.go [hunk 3]
```diff
@@ -98,6 +106,13 @@ func (f *Flags) AddTo(flagSet *pflag.FlagSet) {
 		"Enable leader election for controller manager. Enabling this will"+
 			" ensure there is only one active controller manager.",
 	)
+	_ = flagSet.MarkDeprecated("enable-leader-election", "Use the flag leader-elect instead of.")
```
Comment (estroz) [inline internal/ansible/flags/flag.go:109]:
```suggestion
	_ = flagSet.MarkDeprecated("enable-leader-election", "use --leader-elect instead")
```

File: internal/ansible/flags/flag.go [hunk 4]
```diff
@@ -27,11 +27,13 @@ type Flags struct {
 	WatchesFile             string
 	InjectOwnerRef          bool
 	EnableLeaderElection    bool
```
Comment (estroz) [inline internal/ansible/flags/flag.go:29]: Use the new field in the old flags, it's less confusing.
```suggestion
```

File: internal/ansible/flags/flag.go [hunk 5]
```diff
@@ -27,11 +27,13 @@ type Flags struct {
...
+	LeaderElection          bool
 	MaxConcurrentReconciles int
 	AnsibleVerbosity        int
 	AnsibleRolesPath        string
 	AnsibleCollectionsPath  string
 	MetricsAddress          string
```
Comment (estroz) [inline internal/ansible/flags/flag.go:35]:
```suggestion
```

File: internal/cmd/ansible-operator/run/cmd.go [hunk 6]
```diff
@@ -99,12 +99,26 @@ func run(cmd *cobra.Command, f *flags.Flags) {
 		}
 	}
 
+	// We need to check both because EnableLeaderElection is deprecated
+	// If one of them is true than it means that the default value was changed
+	// todo:remove it for 2.0.0
+	if f.LeaderElection || f.EnableLeaderElection {
+		f.LeaderElection = true
+	}
```
Comment (estroz) [inline internal/cmd/ansible-operator/run/cmd.go:107]: You should error out if both are set instead.
```suggestion
	if cmd.Flags().Changed("leader-elect") && cmd.Flags().Changed("enable-leader-election") {
		return errors.New("only one of --leader-elect and --enable-leader-election may be set")
	}
```

File: internal/plugins/ansible/v1/scaffolds/init.go [hunk 7]
```diff
@@ -97,11 +97,11 @@ func (s *initScaffolder) scaffold() error {
 		&prometheus.Kustomization{},
 		&prometheus.ServiceMonitor{},
 
-		&manager.Manager{Image: imageName},
+		&manager.Config{Image: imageName},
```
Comment (estroz) [inline internal/plugins/ansible/v1/scaffolds/init.go:100]: These changes are unrelated to flag name changes, can you put them in a separate PR?

**Review by estroz — CHANGES_REQUESTED** [review]
A bunch of changes in here have nothing to do with flag name changes, although cosmetic.

File: internal/plugins/ansible/v1/scaffolds/init.go (hunk 7 above)
Comment (camilamacedo86) [reply internal/plugins/ansible/v1/scaffolds/init.go:100]: I need do the changes to be able to compare the files. 
I am just renaming the files and keeping then == upstream for we are able to diff

File: testdata/helm/memcached-operator/config/default/kustomization.yaml [hunk 8]
```diff
@@ -20,7 +20,7 @@ bases:
 - ../prometheus
 
 patchesStrategicMerge:
-  # Protect the /metrics endpoint by putting it behind auth.
-  # If you want your controller-manager to expose the /metrics
-  # endpoint w/o any authn/z, please comment the following line.
+# Protect the /metrics endpoint by putting it behind auth.
+# If you want your controller-manager to expose the /metrics
+# endpoint w/o any authn/z, please comment the following line.
```
Comment (camilamacedo86) [inline testdata/helm/memcached-operator/config/default/kustomization.yaml:25]: these changes are because of the alignment to keep the file as upstream.

File: testdata/ansible/memcached-operator/config/default/kustomization.yaml [hunk 9]
```diff
@@ -20,7 +20,7 @@ bases:
 - ../prometheus
 
 patchesStrategicMerge:
-  # Protect the /metrics endpoint by putting it behind auth.
-  # If you want your controller-manager to expose the /metrics
-  # endpoint w/o any authn/z, please comment the following line.
+# Protect the /metrics endpoint by putting it behind auth.
+# If you want your controller-manager to expose the /metrics
+# endpoint w/o any authn/z, please comment the following line.
```
Comment (camilamacedo86) [inline testdata/ansible/memcached-operator/config/default/kustomization.yaml:25]: these changes are because of the alignment to keep the file upstream.

File: internal/plugins/helm/v1/scaffolds/internal/templates/config/manager/kustomization.go [hunk 10]
```diff
@@ -24,12 +24,12 @@ import (
...
 
-// Kustomization scaffolds the Kustomization file in manager folder.
+// Kustomization scaffolds a file that defines the kustomization scheme for the manager folder
 type Kustomization struct {
 	file.TemplateMixin
 }
 
-// SetTemplateDefaults implements input.Template
+// SetTemplateDefaults implements file.Template
 func (f *Kustomization) SetTemplateDefaults() error {
```
Comment (camilamacedo86) [inline internal/plugins/helm/v1/scaffolds/internal/templates/config/manager/kustomization.go:33]: these changes are because of the alignment to keep the file such as upstream.

File: internal/plugins/helm/v1/scaffolds/internal/templates/config/manager/config.go [hunk 11]
```diff
@@ -23,29 +22,29 @@ import (
...
 	file.ProjectNameMixin
 
 	// Image is controller manager image name
 	Image string
 }
 
-// SetTemplateDefaults implements input.Template
-func (f *Manager) SetTemplateDefaults() error {
+// SetTemplateDefaults implements file.Template
+func (f *Config) SetTemplateDefaults() error {
```
Comment (camilamacedo86) [inline internal/plugins/helm/v1/scaffolds/internal/templates/config/manager/config.go:37]: these changes are because of the alignment to keep the file such as upstream.

File: internal/plugins/helm/v1/scaffolds/internal/templates/config/manager/config.go [hunk 12]
```diff
@@ -23,29 +22,29 @@ import (
 	"sigs.k8s.io/kubebuilder/v3/pkg/model/file"
 )
 
-var _ file.Template = &Manager{}
+var _ file.Template = &Config{}
 
-// Manager scaffolds yaml config for the manager.
-type Manager struct {
+// Config scaffolds a file that defines the namespace and the manager deployment
+type Config struct {
```
Comment (camilamacedo86) [inline internal/plugins/helm/v1/scaffolds/internal/templates/config/manager/config.go:28]: these changes are because of the alignment to keep the file such as upstream.

File: internal/plugins/helm/v1/scaffolds/internal/templates/config/kdefault/manager_auth_proxy_patch.go [hunk 13]
```diff
@@ -23,17 +23,16 @@ import (
...
+// ManagerAuthProxyPatch scaffolds a file that defines the patch that enables prometheus metrics for the manager
+type ManagerAuthProxyPatch struct {
 	file.TemplateMixin
 	file.ProjectNameMixin
 }
 
-// SetTemplateDefaults implements input.Template
-func (f *AuthProxyPatch) SetTemplateDefaults() error {
+// SetTemplateDefaults implements file.Template
+func (f *ManagerAuthProxyPatch) SetTemplateDefaults() error {
```
Comment (camilamacedo86) [inline internal/plugins/helm/v1/scaffolds/internal/templates/config/kdefault/manager_auth_proxy_patch.go:35]: these changes are because of the alignment to keep the file such as upstream.

File: internal/plugins/helm/v1/scaffolds/init.go [hunk 14]
```diff
@@ -94,10 +94,10 @@ func (s *initScaffolder) scaffold() error {
 		&rbac.ManagerRole{},
 		&rbac.ManagerRoleBinding{},
 		&manager.Kustomization{},
-		&manager.Manager{Image: imageName},
+		&manager.Config{Image: imageName},
 		&prometheus.Kustomization{},
 		&prometheus.ServiceMonitor{},
-		&kdefault.AuthProxyPatch{},
+		&kdefault.ManagerAuthProxyPatch{},
```
Comment (camilamacedo86) [inline internal/plugins/helm/v1/scaffolds/init.go:100]: these changes are because of the alignment to keep the file such as upstream.
just replace the names

File: internal/plugins/ansible/v1/scaffolds/internal/templates/config/manager/kustomization.go [hunk 15]
```diff
@@ -24,12 +24,12 @@ import (
...
 
-// Kustomization scaffolds the Kustomization file in manager folder.
+// Kustomization scaffolds a file that defines the kustomization scheme for the manager folder
 type Kustomization struct {
 	file.TemplateMixin
 }
 
-// SetTemplateDefaults implements input.Template
+// SetTemplateDefaults implements file.Template
 func (f *Kustomization) SetTemplateDefaults() error {
```
Comment (camilamacedo86) [inline internal/plugins/ansible/v1/scaffolds/internal/templates/config/manager/kustomization.go:33]: these changes are because of the alignment to keep the file such as upstream.

File: internal/plugins/ansible/v1/scaffolds/internal/templates/config/manager/config.go [hunk 16]
```diff
@@ -23,19 +23,19 @@ import (
 	"sigs.k8s.io/kubebuilder/v3/pkg/model/file"
 )
 
-var _ file.Template = &Manager{}
+var _ file.Template = &Config{}
 
-// Manager scaffolds yaml config for the manager.
-type Manager struct {
+// Config scaffolds a file that defines the namespace and the manager deployment
+type Config struct {
```
Comment (camilamacedo86) [inline internal/plugins/ansible/v1/scaffolds/internal/templates/config/manager/config.go:29]: these changes are because of the alignment to keep the file such as upstream.

**Comment by camilamacedo86** [comment]
Hi @estroz, 

Really thank you. All suggestion (less this nit https://github.com/operator-framework/operator-sdk/pull/4654#discussion_r594718239)  are addressed, PLease, feel free to check.

File: changelog/fragments/ansible-helm-flags.yaml [hunk 17]
```diff
@@ -0,0 +1,71 @@
+# entries is a list of entries to include in
+# release notes and/or the migration guide
+entries:
+  - description: >
+      (ansible/v1)(helm/v1) In order to align the Ansible/Helm based-projects with the Kubernetes nomenclatures the flags `--enable-leader-election` and `--metrics-addr` were deprecated and replaced by `--leader-elect` and `--metrics-bind-address` respectively.
```
Comment (estroz) [inline changelog/fragments/ansible-helm-flags.yaml:5]:
```suggestion
      (ansible/v1, helm/v1) The flags `--enable-leader-election` and `--metrics-addr` were deprecated in favor of `--leader-elect` and `--metrics-bind-address`, respectively, to follow upstream conventions.
```

File: changelog/fragments/ansible-helm-flags.yaml [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+    # message in which this file was originally added.
+    #
+    # What is the pull request number (without the "#")?
+    # pull_request_override: 0
+
+
+    # Migration can be defined to automatically add a section to
+    # the migration guide. This is required for breaking changes.
+    migration:
+      header: (ansible/v1)(helm/v1) Replace the deprecated flags
```
Comment (estroz) [inline changelog/fragments/ansible-helm-flags.yaml:31]:
```suggestion
      header: (ansible/v1, helm/v1) Replace deprecated leader election and metrics address flags
```

File: changelog/fragments/ansible-helm-flags.yaml [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+      body: >
+        Replace the flags `--enable-leader-election` and `--metrics-addr` which were deprecated with `--leader-elect` and `--metrics-bind-address` respectively.
```
Comment (estroz) [inline changelog/fragments/ansible-helm-flags.yaml:33]:
```suggestion
        Replace deprecated flags `--enable-leader-election` and `--metrics-addr` with `--leader-elect` and `--metrics-bind-address`, respectively.
```

File: changelog/fragments/ansible-helm-flags.yaml [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+  - description: >
+      (ansible/v1)(helm/v1) Ensure that the `--health-probe-bind-address` port will be informed to the proxy.
```
Comment (estroz) [inline changelog/fragments/ansible-helm-flags.yaml:35]:
```suggestion
      (ansible/v1, helm/v1) Explicitly set `--health-probe-bind-address` in the manager's auth proxy patch.
```

File: changelog/fragments/ansible-helm-flags.yaml [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    # What is the pull request number (without the "#")?
+    # pull_request_override: 0
+
+
+    # Migration can be defined to automatically add a section to
+    # the migration guide. This is required for breaking changes.
+    migration:
+      header: (ansible/v1)(helm/v1) Ensure that the `--health-probe-bind-address` port will be informed to the proxy.
+      body: >
+        Add the arg `--health-probe-bind-address=:8081` to the `config/default/manager_auth_proxy_patch.yaml`:
```
Comment (estroz) [inline changelog/fragments/ansible-helm-flags.yaml:63]: This is different for ansible (port 6789). Make another changelog entry I guess?

**Review by estroz — CHANGES_REQUESTED** [review]
Changelog nits, otherwise LGTM.

**Comment by camilamacedo86** [comment]
@estroz All done  tks.

**Review by estroz — APPROVED** [review]
/lgtm
